### PR TITLE
routing+htlcswitch: finalize switch of CLTV delta directionality in path finding and link forwarding

### LINKED ISSUE #1398: low CLTV creates routing errors
### Background

Here is a link to the slack convo for reference, but I will summarize below: https://lightningcommunity.slack.com/archives/C6BKD3RKR/p1529106498000126

I had set my `time_lock_delta` setting to 4 (from the 144 default) using `lncli updatechanpolicy`. I do not remember if I was able to successfully make payments with this setting, but I think I was. A few days later I tried to pay `satoshis.place` which I had previously done successfully many times. However, payments began to fail with the following error:
```
{
        "payment_error": "unable to route payment to destination: IncorrectCltvExpiry(expiry=527800, update=(lnwire.ChannelUpdate) {
 Signature: (lnwire.Sig) (len=64 cap=64) {
  00000000  c0 4e ad f5 99 78 70 de  62 47 eb e4 2b 54 03 51  |.N...xp.bG..+T.Q|
  00000010  c3 44 c6 9f 0c 96 c4 66  8d e8 fa c2 ac c3 07 a8  |.D.....f........|
  00000020  12 cc a0 1b 07 bc c5 67  b5 f2 48 5a 72 7d 3c 10  |.......g..HZr}\u003c.|
  00000030  54 35 99 a0 2c c5 68 14  b7 c2 0d 86 ce 2a 7c c7  |T5..,.h......*|.|
 },
 ChainHash: (chainhash.Hash) (len=32 cap=32) 000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f,
 ShortChannelID: (lnwire.ShortChannelID) 527196:849:1,
 Timestamp: (uint32) 1529018851,
 Flags: (lnwire.ChanUpdateFlag) 0,
 TimeLockDelta: (uint16) 4,
 HtlcMinimumMsat: (lnwire.MilliSatoshi) 1000 mSAT,
 BaseFee: (uint32) 0,
 FeeRate: (uint32) 1
}
",
        "payment_preimage": "",
        "payment_route": null
}
```

Upon attempting to pay the same invoice a second time, the following error would occur:

```
{
        "payment_error": "unable to find a path to destination",
        "payment_preimage": "",
        "payment_route": null
}
```

After running `updatechanpolicy` again payments still failed with the "unable to find a path to destination" error for a few minutes and then began to succeed.

### Your environment
`lnd version 0.4.2-beta commit=b5a228808b0446d6758a23e24d6c686aab9936fc`
`Linux 4.4.0-1052-aws #61-Ubuntu SM`
Bitcoind backend

### Steps to reproduce

See description

### Expected behaviour

Payment should occur or understandable error should be reported. 

### Actual behaviour

CLTV error described was given however it seems that this shouldn't have happened given other routes were available but seemingly not attempted.

Read the slack thread for some additional context

### LINKED ISSUE #1431: Link rejects HTLCs that meet CLTV forwarding policy of outbound link but not inbound link
### Background

When a single node's channels have diverse timelock delta forwarding policies (e.g. Channel A has a forwarding policy of 
```
{
  "timeLockDelta":130,
  "minHtlc":"1000",
  "feeBaseMsat":"0",
  "feeRateMilliMsat":"6"
}
```
and Channel B has a forwarding policy of
```
{
  "timeLockDelta":144,
  "minHtlc":"1000",
  "feeBaseMsat":"0",
  "feeRateMilliMsat":"6"
}
```
)

It rejects incoming HTLCs to forward that have enough timelock according to the outbound link's policies but not the inbound link's policies.

For example, if a route was forwarding through the node from Channel B to Channel A, and only had 130 of time lock delta, it would be rejected with the following error message:

```
2018-06-22 00:52:44.183 [ERR] HSWC: Incoming htlc(ec387ebcd20521ae17c81841020c488f7ec9ce10287856cd77cc24bed7633f53) has incorrect time-lock value: expected at least 144 block delta, got 130 block delta
```

### Your environment

* version of `lnd`: `0.4.2` commit `fc658dbf6043951f6ba8b8e86f4e67bc903cf6b7`
* which operating system (`uname -a` on *Nix): Darwin Kernel Version 17.5.0
* version of `btcd`, `bitcoind`, or other backend: btcd commit `bc09449`
* any other relevant environment details

### Steps to reproduce
1. Create 3 Nodes, A, B, and C
2. Open a channel between A -> B, and B -> C
3. Set B's forwarding policy for A->B to a timelock delta of 100 (from the default of 144)
4. Send a payment from A to C.

### Expected behaviour

B should forward the HTLC on the basis of its forwarding policy for B->C, not it's forwarding policy for A->B.

I believe I've found the source of the issue in https://github.com/lightningnetwork/lnd/blob/fc658dbf6043951f6ba8b8e86f4e67bc903cf6b7/htlcswitch/link.go :

Setting the `hopPolicy` from the inbound link
https://github.com/lightningnetwork/lnd/blob/fc658dbf6043951f6ba8b8e86f4e67bc903cf6b7/htlcswitch/link.go#L2347

Using the `timeDelta` from the inbound link's `hopPolicy`
https://github.com/lightningnetwork/lnd/blob/fc658dbf6043951f6ba8b8e86f4e67bc903cf6b7/htlcswitch/link.go#L2353

Using the inbound link's `timeDelta` to check against the outgoing cltv
https://github.com/lightningnetwork/lnd/blob/fc658dbf6043951f6ba8b8e86f4e67bc903cf6b7/htlcswitch/link.go#L2354
https://github.com/lightningnetwork/lnd/blob/fc658dbf6043951f6ba8b8e86f4e67bc903cf6b7/htlcswitch/link.go#L2386

This is similar to the fee issue that was fixed in #1019 wherein the fee was being measured against the incoming link's policy instead of the outgoing link. I suspect that since most nodes on the network use the same CLTV delta policy on all of their channels, this issue hasn't surfaced as broadly on mainnet.

This behavioral expectation also comes from the spec: https://github.com/lightningnetwork/lightning-rfc/blob/master/07-routing-gossip.md#the-channel_update-message

Of course, if this is a misunderstanding of the spec, let me know.

### Actual behaviour

The forwarding node fails to forward with this error message:
```
2018-06-22 00:52:44.183 [ERR] HSWC: Incoming htlc(ec387ebcd20521ae17c81841020c488f7ec9ce10287856cd77cc24bed7633f53) has incorrect time-lock value: expected at least 144 block delta, got 130 block delta
```

The sending node receives a failed payment with:
`Attempt to send payment ec387ebcd20521ae17c81841020c488f7ec9ce10287856cd77cc24bed7633f53 failed: IncorrectCltvExpiry`

### PR DESCRIPTION
**Pull request by Roasbeef**
In this commit, we fix an old-ish bug within the project. This bug was first introduced due to an incomplete migration to the swapped directionality of the CLTV value within the `ChannelUpdate` message. Now the value from A -> B is B's policy, while in the past, this value was swapped.

Within the `newRoute` method in the router, we failed to carry this over, and were still using the CLTV delta of the current hop, rather than that of the prior hop (the one later in the route). Additionally, the expiry values within the spec json example were also incorrect, causing this issue to go undetected until now. 

This "failure to propagate the CLTV direction swap" carried over to the `htlcswitch` package where the `generateHops` method had a similar bug which caused it to generate an invalid payload. Within the link before this PR, we would actually apply the time lock policy verification at the incoming link instead of the outgoing link. As a result, if a node had two links with distinct CLTV policies (p1.delta > p2.delta), all HTLC's routing in the direction of the greater delta would fail as the link was erroneously applying the incoming policy instead of the outgoing policy. 

In order to fix the above bug, we've extended the `HtlcSatifiesPolicy` method to also factor in the time lock information for that HTLC. We've also added a test which fails without the set of changes included in this PR. 

We should follow up this PR with a change to unify the hop payload generation logic across the `routing` and `htlcswitch` packages. This way, future fixes/modifications can be made in a single location, rather than across to packages. 

One last issue still linger: we don't store the min cltv delta used for a particular invoices on disk. As a result, atm, we always just use our current routing policy which is incorrect if the user ever specifies a manual final cltv value that's lower than the value of the accepting link.

Fixes #1431 .
Fixes #1398.

### CODE REVIEW

File: htlcswitch/link_test.go [hunk 1]
```diff
@@ -4527,3 +4528,56 @@ func TestExpectedFee(t *testing.T) {
...
+	channels, cleanUp, _, err := createClusterChannels(
+		btcutil.SatoshiPerBitcoin*3,
+		btcutil.SatoshiPerBitcoin*5,
+	)
+	if err != nil {
+		t.Fatalf("unable to create channel: %v", err)
+	}
+	defer cleanUp()
+
+	fmt.Println(testStartingHeight)
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply htlcswitch/link_test.go]: Fixed!

File: htlcswitch/test_utils.go [hunk 2]
```diff
@@ -601,15 +601,17 @@ func generateHops(payAmt lnwire.MilliSatoshi, startingHeight uint32,
...
-		if i != len(path)-1 {
-			delta := path[i].cfg.FwrdingPolicy.TimeLockDelta
+		if i == len(path)-1 {
+			totalTimelock += lastHop.cfg.FwrdingPolicy.TimeLockDelta
+			timeLock = totalTimelock
+		} else {
+			// Otherwise, the outgoing time lock should be the
+			// incoming timelock minus their specified delta.
+			delta := path[i+1].cfg.FwrdingPolicy.TimeLockDelta
+			totalTimelock += delta
```
Comment (cfromknecht) [inline htlcswitch/test_utils.go]: order of operations here is a little confusing, is this equivalent to
```golang
timeLock = totalTimeLock
totalTimeLock += path[i+1].cfg.FwrdingPolicy.TimeLockDelta
```
?

File: htlcswitch/link.go [hunk 3]
```diff
@@ -1802,6 +1802,54 @@ func (l *channelLink) HtlcSatifiesPolicy(payHash [32]byte,
...
+
+		return failure
+	}
+
+	// Finally, we'll ensure that the time-lock on the outgoing HTLC meets
+	// the following constraint: the incoming time-lock minus our time-lock
+	// delta should equal the outgoing time lock. Otherwise, whether the
+	// sender messed up, or an intermediate node tampered with the HTLC.
+	if incomingTimeout-timeDelta < outgoingTimeout {
+		log.Errorf("Incoming htlc(%x) has incorrect time-lock value: "+
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply htlcswitch/link.go]: Fixed.

File: htlcswitch/test_utils.go (hunk 2 above)
Comment (Roasbeef) [reply htlcswitch/test_utils.go]: Yep it's the same, but unrolled in order to stress that: outgoingLock = incomingLock - delta

File: htlcswitch/interfaces.go [hunk 4]
```diff
@@ -81,8 +81,10 @@ type ChannelLink interface {
 	// Otherwise, a valid protocol failure message should be returned in
 	// order to signal to the source of the HTLC, the policy consistency
 	// issue.
-	HtlcSatifiesPolicy(payHash [32]byte,
-		incomingAmt, amtToForward lnwire.MilliSatoshi) lnwire.FailureMessage
+	HtlcSatifiesPolicy(payHash [32]byte, incomingAmt lnwire.MilliSatoshi,
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply htlcswitch/interfaces.go]: Hmm, yeh agree we can pass in a struct instead here as the set of args is starting to grow a bit. Would prefer to do it in a follow up, as the `htlcPacket` struct itself could be streamlined a bit by just embedding the `HopPayload` struct for example.

**Pull request merged**